libobs/util/pipe-posix.c

```diff
@@ -16,34 +16,108 @@
 
 #include <stdio.h>
 #include <sys/wait.h>
+#include <unistd.h>
+#include <errno.h>
+#include <spawn.h>
 
 #include "bmem.h"
 #include "pipe.h"
 
 struct os_process_pipe {
 	bool read_pipe;
+	int pid;
 	FILE *file;
+	FILE *err_file;
 };
 
 os_process_pipe_t *os_process_pipe_create(const char *cmd_line,
 					  const char *type)
 {
-	struct os_process_pipe pipe = {0};
+	struct os_process_pipe process_pipe = {0};
 	struct os_process_pipe *out;
+	posix_spawn_file_actions_t file_actions;
 
 	if (!cmd_line || !type) {
 		return NULL;
 	}
 
-	pipe.file = popen(cmd_line, type);
-	pipe.read_pipe = *type == 'r';
+	process_pipe.read_pipe = *type == 'r';
+
+	int mainfds[2] = {0};
+	int errfds[2] = {0};
+
+	if (pipe(mainfds) != 0) {
+		return NULL;
+	}
+
+	if (pipe(errfds) != 0) {
+		close(mainfds[0]);
+		close(mainfds[1]);
+
+		return NULL;
+	}
+
+	if (posix_spawn_file_actions_init(&file_actions) != 0) {
+		close(mainfds[0]);
+		close(mainfds[1]);
+		close(errfds[0]);
+		close(errfds[1]);
+
+		return NULL;
+	}
+
+	if (process_pipe.read_pipe) {
+		posix_spawn_file_actions_addclose(&file_actions, mainfds[0]);
+		if (mainfds[1] != STDOUT_FILENO) {
+			posix_spawn_file_actions_adddup2(
+				&file_actions, mainfds[1], STDOUT_FILENO);
+			posix_spawn_file_actions_addclose(&file_actions,
+							  mainfds[0]);
+		}
+	} else {
+		if (mainfds[0] != STDIN_FILENO) {
+			posix_spawn_file_actions_adddup2(
+				&file_actions, mainfds[0], STDIN_FILENO);
+			posix_spawn_file_actions_addclose(&file_actions,
+							  mainfds[1]);
+		}
+	}
+
+	posix_spawn_file_actions_addclose(&file_actions, errfds[0]);
+	posix_spawn_file_actions_adddup2(&file_actions, errfds[1],
+					 STDERR_FILENO);
+
+	int pid;
+	char *argv[4] = {"sh", "-c", (char *)cmd_line, NULL};
+
+	int ret = posix_spawn(&pid, "/bin/sh", &file_actions, NULL, argv, NULL);
```

Comment:
Instead of letting `sh` to parse the command line, can we add a new limbos API that takes an array of arguments?

Reply:
I intend to do that, but Windows doesn't actually support that natively so we have to do some extra work there. I wanted this PR to get merged first to base my changes on to avoid dealing with merge conflicts. We need to keep this method for ABI compatibility anyway (at least for a while)